lib/collections.c

```diff
@@ -287,9 +285,40 @@ Dictionary* dictionary_init(void)
     return dictionary;
 }
 
-static void dictionary_rebuild(Dictionary* dictionary)
+static uint64_t dictionary_lookup(Dictionary* dictionary, const char* key)
+{
+    uint64_t index = hash_key(key) % dictionary->max_length;
+    while (dictionary->key_value_pairs[index].key != NULL)
+    {
+        if (strcmp(dictionary->key_value_pairs[index].key, key) == 0)
+        {
+            return index;
+        }
+        index = (index + 1) % dictionary->max_length;
+    }
+    return -1;
+}
+
+bool dictionary_contains(Dictionary* dictionary, const char* key)
+{
+    uint64_t index = hash_key(key) % dictionary->max_length;
+    while (dictionary->key_value_pairs[index].key != NULL)
+    {
+        if (strcmp(dictionary->key_value_pairs[index].key, key) == 0)
+        {
+            return true;
+        }
+        index = (index + 1) % dictionary->max_length;
+    }
+    return false;
+}
+
+void dictionary_set(Dictionary* dictionary, const char* key, void* value);
+
+static void dictionary_rebuild(Dictionary** dictionary_ptr)
```

Comment:
This is pretty subtle, but I think having a `Dictionary**` in a helper method while the main API takes `Dictionary*` won't work.

---

If a user of the library has a local variable `Dictionary *d` pointing to an existing dict, and they call a method that resizes (say, `remove`), they're calling it as `dict_remove(d)`.

This means that no matter what happens, the local variable in that user code will stay pointing to the same `Dictionary`. But in the implementation of `remove` when you `rebuild`, you're modifying the `Dictionary*` *local variable of rebuild* to point to the new `Dictionary` -- not the caller's local variable.

Hopefully that makes sense, but no worries if it doesn't; this explanation is probably a bit incomprehensible without a diagram ^^'

lmk what you think

---

And as far as what to do instead, I also have some thoughts, so lmk if you'd like to hear them!

Reply:
Hmmmmm...let's definitely chat through this on Wednesday or Thursday live. (Will hold off on making any substantive changes until we can sync IRL.)

Reply:
Sure thing!

I'll do the same. Still planning to add nice docs to things, for ownership contracts and all that -- but I'm hoping to avoid merge conflicts by waiting until this change lands on main.